### PR TITLE
Basic support for xiangshan

### PR DESCRIPTION
**Pull request by ljwljwljwljw**
TODO:

* Change `xiangshan` to `yanqihu` and `nanhu`

### CODE REVIEW

File: gcc/config/riscv/riscv.cc [hunk 1]
```diff
@@ -208,6 +208,19 @@ struct riscv_integer_op {
...
+  RISCV_FUSE_ZEXTW = (1 << 0),
+  RISCV_FUSE_ZEXTH = (1 << 1),
+  RISCV_FUSE_ZEXTWS = (1 << 2),
+  RISCV_FUSE_LDINDEXED = (1 << 3),
+  RISCV_FUSE_LUI_ADDI = (1 << 4),
+  RISCV_FUSE_AUIPC_ADDI = (1 << 5),
+  RISCV_FUSE_LUI_LD = (1 << 6),
+  RISCV_FUSE_AUIPC_LD = (1 << 7),
+};
+
```
Comment (pz9115) [inline gcc/config/riscv/riscv.cc:223]: This need wait [upstream](https://gcc.gnu.org/pipermail/gcc-patches/2021-November/584755.html) decide whether to keep the fusion function. We will talk about this on next RISC-V GNU toolchain meetings

**Review by pz9115 — APPROVED** [review]
Except the fusion function problem, this patch is well done.

**Review by Liaoshihua — COMMENTED** [review]
You need add testcase for  "nanhu" or "xiangshan", the example is 

[gcc\testsuite\gcc.target\riscv\mcpu-1.c](https://github.com/plctlab/riscv-gcc/blob/xiangshan/gcc/testsuite/gcc.target/riscv/mcpu-1.c)

```
/* { dg-do compile } */
/* { dg-skip-if "-march given" { *-*-* } { "-march=*" } } */
/* { dg-options "-mcpu=sifive-e20 -mabi=ilp32" } */
/* sifive-e20 = rv32imc */

#if !((__riscv_xlen == 32)		\
      && !defined(__riscv_32e)		\
      && defined(__riscv_mul)		\
      && !defined(__riscv_atomic)	\
      && !defined(__riscv_flen)		\
      && defined(__riscv_compressed))
#error "unexpected arch"
#endif

int main()
{
  return 0;
}
```

You need modify  `/* { dg-options "-mcpu=sifive-e20 -mabi=ilp32" } */` to your mcpu

**Comment by pz9115** [comment]
The upstream has supported basic instruction fusion frame in 
https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=c177f28d601408180fdb2db0d5ba89d53b370b5e
so merge this PR and ready to rebase to a upstream version.

**Pull request merged**